keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,970 @@
+is ongoing - no indication is possible to show based on CRI itself. At least
+containerd runtime is already providing such information based on the
+`ctr image pull docker.io/library/busybox:latest`.
+
+#### Story 3
+
+Particular registry is blocked by a firewall in the organization where the Kubernetes cluster is
+running, and workload is trying to use an image that is impossible to download without explicit
+network error. Pod in such situation will be pending the image download for a long time, which can
+be shortened to a no-progress timeout set in ImagePullWithProgress call.
+
+### Notes/Constraints/Caveats (Optional)
+
+It is fairly impossible to get progress about unpacking the image layers, therefore for
+- kubelet:  this operation can be silent without progress reports
+- command line tools - stage of unpacking can be reported with verbose option in request
+
+In case of lazy pulling, which allows starting the container while its image not yet fully downloaded,
+the minimum amount of data the runtime needs to pull (as opposed to total / complete image size)
+in order to start the container should be considered the total amount of download being in progress.
+For instance, if the complete container image consists of 10 layers, 1 GiB per layer, and runtime
+can start the container with only 4 layers pulled, then 4GiB should be considered the total download
+amount and the progress message to Kubelet at hypothetical 1GiB progres point would look like:
+```
+{
+  image_ref: "registryA.com/repo/imageB:tagC",
+  offset: 1073741824,
+  total: 4294967296,
+}
+```
+When downloading the minimum needed layers to start the container is done, the progress reports
+stream can be closed and pull operation can proceed silently in the runtime without notifications
+sent to Kubelet for the rest of the image being lazily pulled as needed.
+
+If ImagePullProgress is enabled for kubelet but runtime does not support it, the fallback should
+use regular silent ImagePull API call.
+
+### Risks and Mitigations
+
+The risks are minimal, introduction of new API is not affecting any other APIs.
+
+Misconfiguration of image pull request granularity for kubelet can result in big number
+of progress responses published by runtime into return stream. This can be mitigated by kubelet
+throttling the amount of events it publishes to Pod object (publishing less than received from
+runtime), as well as in runtime by metering amount of responses being sent within time interval.
+Intention of this KEP's CRI extension is to provide not-so-frequent updates for Kubelet (e.g. once
+a minute), while having a possibility to have more-frequent updates for cli tools (e.g. once a
+second).
+
+Runtime can be such that does not support image pull progress reporting. In this case fallback to
+regular image pulling call should happen on client side (kubelet, cli tool, other entities
+requesting image to be pulled from runtime).
+
```

Comment:
Performance will be impacted. There should be an effort to identify the overhead and inform via docs.

Reply:
Added note about updating docs.